### PR TITLE
DCOS_OSS-5083 Use PATH for coreutils and procps bins

### PR DESCRIPTION
**Pull request by wolf31o2**
Rather than supply an explicit path to the binaries provided by the
coreutils or procps packages, fetch them from the PATH. These are
binaries such as `cp` or `kill` and tend to be assumed to be in the
default PATH on any system. However, these are different prefixes on
RHEL-based systems and Debian-based systems. Rather than be explicit on
the prefix, let's assume the PATH includes these items. This will always
be true on Redhat and Ubuntu installations.

Signed-off-by: Chris Gianelloni <cgianelloni@applause.com>

## High-level description

Removes explicit PATHs on coreutils or procps binaries so we can install/run on Ubuntu.

## Corresponding DC/OS tickets (obligatory)

  - [DCOS_OSS-5083](https://jira.mesosphere.com/browse/DCOS_OSS-5083) Use common tools from PATH versus explicit paths

## Checklist for all PRs

  - [ ] Added a comprehensible changelog entry to `CHANGES.md` or explain why this is not a user-facing change:
  - [ ] Included a test which will fail if code is reverted but test is not. If there is no test please explain here:
  - [x] Read the [DC/OS contributing guidelines](https://github.com/dcos/dcos/blob/master/contributing.md)
  - [x] Followed relevant code rules [Rules for Packages and Systemd](https://github.com/dcos/dcos/tree/master/docs)

## Checklist for component/package updates:

If you are changing components or packages in DC/OS (e.g. you are bumping the sha or ref of anything underneath `packages`), then in addition to the above please also include:

  - [ ] Change log from the last version integrated (this should be a link to commits for easy verification and review): [example](https://github.com/dcos/dcos-mesos-modules/compare/f6fa27d7c40f4207ba3bb2274e2cfe79b62a395a...6660b90fbbf69a15ef46d0184e36755881d6a5ae)
  - [ ] Test Results: [link to CI job test results for component]
  - [ ] Code Coverage (if available): [link to code coverage report]
___
**PLEASE FILL IN THE TEMPLATE ABOVE** / **DO NOT REMOVE ANY SECTIONS ABOVE THIS LINE**

## Instructions and review process

**What is the review process and when will my changes land?**

All PRs require 2 approvals using GitHub's [pull request reviews](https://help.github.com/articles/about-pull-request-reviews/).

Reviewers should be:
* Developers who understand the code being modified.
* Developers responsible for code that interacts with or depends on the code being modified.

It is best to proactively ask for 2 reviews by @mentioning the candidate reviewers in the PR comments area. The responsibility is on the developer submitting the PR to follow-up with reviewers and make sure a PR is reviewed in a timely manner. Once a PR has **2 ship-it's**, **no red reviews**, and **all tests are green** it will be included in the [next train](https://github.com/dcos/dcos/blob/master/contributing.md).

### CODE REVIEW

File: gen/dcos-config.yaml [hunk 1]
```diff
@@ -1425,7 +1425,7 @@ package:
     content: |
       {
         "PATH": "/usr/bin:/bin",
-        "SHELL": "/usr/bin/bash",
+        "SHELL": "/bin/bash",
```
Comment (jongiddy) [inline gen/dcos-config.yaml]: Why are you confident `bash` will be available in `/bin` here, when in `gen/azure/cloud-config.yaml` you have used `/usr/bin/env bash`?

File: gen/azure/cloud-config.yaml [hunk 2]
```diff
@@ -11,7 +11,7 @@ root:
       StandardOutput=journal+console
       StandardError=journal+console
       ExecStartPre=/usr/bin/curl -fLsSv --retry 20 -Y 100000 -y 60 -o /var/tmp/d.deb https://download.docker.com/linux/ubuntu/dists/xenial/pool/stable/amd64/docker-ce_17.09.0~ce-0~ubuntu_amd64.deb
-      ExecStart=/usr/bin/bash -c "try=1;until dpkg -D3 -i /var/tmp/d.deb || ((try>9));do echo retry $((try++));sleep $((try*try));done;systemctl --now start docker;systemctl restart docker.socket"
+      ExecStart=/usr/bin/env bash -c "try=1;until dpkg -D3 -i /var/tmp/d.deb || ((try>9));do echo retry $((try++));sleep $((try*try));done;systemctl --now start docker;systemctl restart docker.socket"
```
Comment (jongiddy) [inline gen/azure/cloud-config.yaml]: Since you use relative paths for the other `ExecStart` lines, why use a specific path here for `env`?

**Comment by jgehrcke** [comment]
Tanks @wolf31o2! I have asked around internally towards getting this some reviews. Sorry for the delay so far.

File: gen/azure/cloud-config.yaml (hunk 2 above)
Comment (wolf31o2) [reply gen/azure/cloud-config.yaml]: `env` isn't necessary, at all... I'll give this another pass later today, as I realize there's no reason to touch the provider-specific configs such as this one

**Comment by wolf31o2** [comment]
I'm going to give this another pass. The intent was to generalize the Ubuntu changes in the Azure cloud config so DC/OS can run on stock Ubuntu, without symlinking everything.

**Comment by orsenthil** [comment]
Hi @wolf31o2  - Thanks for your contribution.

I noticed failures like this:

> 2019-04-30 14:39:18 ERROR    dcos_e2e._common | Failed to start dcos-link-env.service: Unit is not loaded properly: Invalid argument.

It means that the changes being introduced are adversely affecting the start of systemd units.

We also have note that, our default testing environment is Cent OS 7.4, and we also support CoreOS. So, changes will have to be compatible with CentOs and CoreOS. 

Thanks, again!

**Comment by orsenthil** [comment]
Please ping us for reviews when you are ready!

**Comment by surdy** [comment]
My understanding of Systemd is that it need absolute paths for the binaries. Even if it is for binaries that would be standard linux binaries found in PATH of the user shell. 
Perhaps my understanding is outdated.  @wolf31o2 did you test the changes you made? Last time I tiried to do this systemd failed to recognize the unit as  valid unless I used absolute path in ExecStart[*] lines

**Comment by surdy** [comment]
Ohh looks like SystemD 239 removed that requirement : https://github.com/systemd/systemd/blob/master/NEWS#L1085
But RHEL/Centos 7.x are stuck on systemd 219 and Container Linux has had exactly one stable release with systemd > 239. We have to support them for the near future based on https://docs.mesosphere.com/version-policy/#dcos-platform-version-compatibility-matrix

Based on this I'm going to decline the change. Hopefully in a brighter future when EL (and all supported distros for DC/OS) has moved to newer Systemd. I would be happy to approve a PR that simplifies systemd units.

File: pkgpanda/__init__.py [hunk 3]
```diff
@@ -612,8 +612,12 @@ def add_user(self, username, groupname):
                                   "automatic user addition is disabled".format(username))
 
         # Add the user:
+        useradd_cmd = 'useradd'
+        if os.path.exists('/usr/sbin/useradd'):
+            useradd_cmd = '/usr/sbin/useradd'
+
         add_user_cmd = [
-            'useradd',
+            useradd_cmd,
```
Comment (branden) [inline pkgpanda/__init__.py]: Why is this change necessary? Is there a competing `useradd` earlier in the `PATH` that we need to avoid?

**Comment by surdy** [comment]
Great point @branden `/bin/bash -c '<cmd>'` might be a great workaround for what we are doing with those binaries.

File: gen/dcos-config.yaml (hunk 1 above)
Comment (wolf31o2) [reply gen/dcos-config.yaml]: I prefer to use `/usr/bin/env bash` in cross-distribution scripts. However, you can't use that for a shell.

```
# CentOS defaults to /usr/bin/bash by PATH
cgianelloni@chris:~/Repos/mine/dcos$ docker run --rm -ti centos:7 bash -c 'type -P bash'
/usr/bin/bash
# However, it's also available as /bin/bash
cgianelloni@chris:~/Repos/mine/dcos$ docker run --rm -ti centos:7 bash -c 'ls -l /bin/bash'
-rwxr-xr-x 1 root root 964608 Oct 30  2018 /bin/bash
# And it's /bin/bash on Ubuntu
cgianelloni@chris:~/Repos/mine/dcos$ docker run --rm -ti ubuntu:18.04 bash -c 'type -P bash'
/bin/bash
```

File: pkgpanda/__init__.py (hunk 3 above)
Comment (wolf31o2) [reply pkgpanda/__init__.py]: I'll add a comment. This is because the PATH used by pkgpanda doesn't include `/usr/sbin` which is the location on Ubuntu. On Azure, it's symlinked into `/usr/bin`, but I was attempting to get this working on stock Ubuntu without customization.

**Comment by wolf31o2** [comment]
I'm going to use `/bin/bash -c` where appropriate, as `/bin/bash` exists on CentOS, CoreOS, and Ubuntu.

**Comment by wolf31o2** [comment]
It looks like either `/bin/bash -c <command>` or `/usr/bin/env <command>` would be appropriate. Any preference?

```
$ docker run --rm -ti centos:7 bash -c 'ls -l /usr/bin/env'
-rwxr-xr-x 1 root root 29048 Oct 30  2018 /usr/bin/env
$ docker run --rm -ti ubuntu:18.04 bash -c 'ls -l /usr/bin/env'
-rwxr-xr-x 1 root root 35000 Jan 18  2018 /usr/bin/env
```

**Comment by surdy** [comment]
@wolf31o2 I have not used either of them in a real world scenario, so I am not aware of limitation/advantages of one over other. However since these are all simple commands (mkdir, ping, sleep etc) I could not think of any corner cases.
@branden do you have any preferences ?
